### ISSUE TITLE
Feature: Async

### ISSUE THREAD

**Issue by lostmsu** [dokan-dev/dokan-dotnet#354]
As in the title - it would be nice to modernize the interface to be asynchronous, and perhaps also switch to `Memory<byte>` or `Span<byte>` for buffers.

A few other performance improvements like iterative file search as mentioned by @LTRData would be welcome too.

**Comment by LTRData** [comment]
I am working on that in a fork here:
https://github.com/LTRData/dokan-dotnet/

The goal is to eventually merge it here as a pull request/branch in one way or another too. But there are lots of breaking changes and changes to implementation interfaces, so it requires some work to migrate existing implementations to it. The performance improvements are so big, I really think it is worth the breaking changes though. We could probably provide some kind of compatibility layer to make migration easier too.

**Comment by LTRData** [comment]
But there is really not anything that needs to be changed to support async as I see it. There should be no issues with calling async code from file system implementation routines. I have done that in lots of implementations without any issues.

**Comment by Liryna** [comment]
@LTRData I totally approve what you said https://github.com/dokan-dev/dokan-dotnet/issues/354#issuecomment-2027829167 Feel free to make the change when you feel comfortable. I agree that keeping the compatibility would be nice.

**Comment by d2dyno1** [comment]
> The performance improvements are so big, I really think it is worth the breaking changes though.

@LTRData Do you have any performance benchmarks for the changes? 🙂

**Comment by LTRData** [comment]
The benchmarks were a bit complicated to make, because they needed two applications. One that made file requests and one Dokan implementation that was monitored by the benchmark. I am thinking of recreating it in a way that I could share more easily.

**Comment by LTRData** [comment]
Some basic performance profiling. Mounted Mirror sample and copied a directory with lots of subdirectories and about 6,000 files out of that mirrored drive to another place.

First, current version. There are lots of allocations of temporary objects that GC collects and frees over time, but it consumes lots of CPU and memory just to do these tasks.
![Image](https://github.com/user-attachments/assets/6c5b530e-83ab-4ee4-b38c-121d7b7f8c84)

Then, current version with `-unsafe`. This avoids some byte arrays and we can clearly see that the pressure on GC and memory consumed is lower.
![Image](https://github.com/user-attachments/assets/04c5e8a9-d86e-4fd7-9959-b13362b7cc70)

Then, my new design without creation of string objects on each request, without allocating byte arrays, with `DokanFileInfo` as a struct instead of object etc:
![Image](https://github.com/user-attachments/assets/fb674428-a9ed-4bbc-a300-de30c293bf62)

My idea now is to create subdirectory in the library and put my modified files there and let the existing files provide the existing API for compatibility. There are some changes needed to implementations for consuming this redesigned API, so we cannot simply replace the old API.

**Comment by lostmsu** [comment]
@LTRData was there any change in practical performance though?

**Comment by LTRData** [comment]
> @LTRData was there any change in practical performance though? As in time to copy.

Not that much, really. It gets a bit more responsive when opening and working with files on Dokan file systems.

But the most notable performance boost come from the lower memory and CPU pressure in scenarios where that is critical. For example, I am working on an application that is used on VDI servers with lots of simultaneous user sessions where each user mounts several different Dokan file systems for accessing cloud file storage etc. There can be hundreds of Dokan file systems on such servers and optimization in ways like this have great impact.

But, please suggest more ideas for measuring performance differences! File copying is not really the most interesting thing to try for all use cases!

**Comment by LTRData** [comment]
Bringing up this topic again, I have just made a branch that includes this implementation. Please make suggestions, ideas and post questions about how it works!
Discussion: https://github.com/dokan-dev/dokan-dotnet/discussions/373
Branch: https://github.com/dokan-dev/dokan-dotnet/tree/IDokanOperations2

**Comment by lostmsu** [comment]
@LTRData I see the use of `Span`-like structures, but no indication of async.

**Comment by LTRData** [comment]
> @LTRData I see the use of `Span`-like structures, but no indication of async.

No, like said earlier in this discussion, the way the library works currently with a thread pool in native code means that a thread needs to wait anyway. Therefore, adding the memory and CPU overhead for async support is pointless and only wastes resources for building state machines etc. If you want to consume async within I/O implementations, you can just do that and wait for result.

The most important thing for performance at this point was to avoid unnecessary heap allocations and avoid GC pressure because many projects that I was working on that used Dokan really needed to optimization and a design change to support that.

A long-term solution for true async would need an async version of the native library with support for completion callbacks instead. That needs to be done first, before implementing it in.NET.

**Comment by Liryna** [comment]
Thanks @LTRData for the massive improvements and (C#) refresh! `DokanOperationsAdapter` is great to have 👍 
Since this is massive I think unless anyone has a blocker in the next few days we can merge it and then see feedback when people start using it.

@LTRData Could you sync at head to have the CI run on the branch ? https://github.com/dokan-dev/dokan-dotnet/commit/a1e4a63efbfd8c3c06b77463616e4e3b49b1457e

**Comment by LTRData** [comment]
> Thanks [@LTRData](https://github.com/LTRData) for the massive improvements and (C#) refresh! `DokanOperationsAdapter` is great to have 👍 Since this is massive I think unless anyone has a blocker in the next few days we can merge it and then see feedback when people start using it.
> 
> [@LTRData](https://github.com/LTRData) Could you sync at head to have the CI run on the branch ? [a1e4a63](https://github.com/dokan-dev/dokan-dotnet/commit/a1e4a63efbfd8c3c06b77463616e4e3b49b1457e)

That made me realize I need to sort out a few things with dependencies etc. I'll fix that!

**Comment by LTRData** [comment]
Done. CI runs seem to have completed successfully now.

**Comment by LTRData** [comment]
> Thanks for the quick fix @LTRData ! Will wait for a few days for anyone to raise their voice and then will merge it.
> Not now but should we one day mark `IDokanOperations` as deprecated then after a time remove it or there is some advantage to keep it ? (Like there are cases where `IDokanOperations` can only be used but not `IDokanOperations2`)

I am not sure about what to do with the old `IDokanOperations` interface. If it does not become to complicated to continue to support it at some point, I guess it makes sense to keep it. It might take some time and effort for implementations to migrate to the new interface.

Maybe not mark it as depreciated in the next version, but the version after it?

**Comment by LTRData** [comment]
Also, regarding the still missing async support, just wanted to reference the old discussion/PR with an idea for an async native library implementation: https://github.com/dokan-dev/dokany/pull/307

Something like that needs to be implemented in the native library first. After that, it is fairly easy to implement async support in the .NET library.

**Comment by Liryna** [comment]
The changes were merged and a new version was released! 🥳  https://github.com/dokan-dev/dokan-dotnet/releases/tag/v2.3.0.1
Please test and use the new `IDokanOperations2` interface and share any feedback 🕶
